src/Service/ElementService.php

```diff
@@ -39,4 +66,299 @@ public static function getCustomViewById(string $id): ?array
+    ): void
+    {
+        $hasChildren = $element->getDao()->hasChildren($user);
+
+        $permissions = $element->getUserPermissions($user);
+
+        $tmpNode['permissions'] = [
+            'remove' => $permissions['delete'],
+            'settings' => $permissions['settings'],
+            'rename' => $permissions['rename'],
+            'publish' => $permissions['publish'],
+            'view' => $permissions['view'],
+            'list' => $permissions['list'],
+        ];
+
+        if ($element instanceof Asset\Folder) {
+            $tmpNode['leaf'] = false;
+            $tmpNode['expanded'] = !$hasChildren;
+            $tmpNode['loaded'] = !$hasChildren;
+            $tmpNode['permissions']['create'] = $permissions['create'];
+            $tmpNode['thumbnail'] = $this->getThumbnailUrl($element, ['origin' => 'treeNode']);
+        } else {
+            $tmpNode['leaf'] = true;
+            $tmpNode['expandable'] = false;
+            $tmpNode['expanded'] = false;
+        }
+
+        $this->assignAssetThumbnailConfig($element, $tmpNode);
+    }
+
+    /**
+     * @throws \Exception
+     */
+    public function assignDataObjectTreeConfig(
+        DataObject\AbstractObject $element,
+        array &$tmpNode,
+        UserProxy|User|null $user
+    ): void
+    {
+        $allowedTypes = [DataObject::OBJECT_TYPE_OBJECT, DataObject::OBJECT_TYPE_FOLDER];
+        if ($element instanceof DataObject\Concrete && $element->getClass()->getShowVariants()) {
+            $allowedTypes[] = DataObject::OBJECT_TYPE_VARIANT;
+        }
+        $hasChildren = $element->getDao()->hasChildren($allowedTypes, null, $user);
+
+        $tmpNode['allowDrop'] = ($tmpNode['type'] ?? false) != DataObject::OBJECT_TYPE_VARIANT;
+        $tmpNode['isTarget'] = true;
+        $tmpNode['allowChildren'] = true;
+        $tmpNode['leaf'] = !$hasChildren;
+        $tmpNode['cls'] = 'pimcore_class_icon ';
+
+        if ($element instanceof DataObject\Concrete) {
+            $tmpNode['published'] = $element->isPublished();
+            $tmpNode['className'] = $element->getClass()->getName();
+
+            if (!$element->isPublished()) {
+                $tmpNode['cls'] .= 'pimcore_unpublished ';
+            }
+
+            $tmpNode['allowVariants'] = $element->getClass()->getAllowVariants();
+        }
+
+        $tmpNode['expanded'] = !$hasChildren;
+        $tmpNode['permissions'] = $element->getUserPermissions($user);
+
+        if ($tmpNode['leaf']) {
+            $tmpNode['expandable'] = false;
+            $tmpNode['leaf'] = false; //this is required to allow drag&drop
+            $tmpNode['expanded'] = true;
+            $tmpNode['loaded'] = true;
+        }
+    }
+
+    /**
+     * @throws \Exception
+     */
+    public function assignDocumentTreeConfig(
+        Document $element,
+        array &$tmpNode,
+        UserProxy|User|null $user
+    ): void
+    {
+        $hasChildren = $element->getDao()->hasChildren(null, Admin::getCurrentUser());
```

Comment:
I have some doubt on having the `$user` parameter passed around by all these functions, probably would be more easier to refactor and just use `Admin::getCurrentUser()`

Reply:
Replaced it with [user loader](https://github.com/pimcore/admin-ui-classic-bundle/pull/129/commits/051bec17723cea0171d282d48fd4af244734c352)